codec-http/src/test/java/io/netty/handler/codec/http/multipart/DefaultHttpDataFactoryTest.java

```diff
@@ -0,0 +1,130 @@
+
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.DefaultHttpRequest;
+import io.netty.handler.codec.http.HttpRequest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static io.netty.handler.codec.http.HttpHeaderValues.IDENTITY;
+import static io.netty.handler.codec.http.HttpMethod.POST;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+import static io.netty.handler.codec.http.multipart.HttpPostBodyUtil.DEFAULT_TEXT_CONTENT_TYPE;
+import static io.netty.util.CharsetUtil.UTF_8;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+public class DefaultHttpDataFactoryTest {
+    // req1 equals req2
+    private static final HttpRequest req1 = new DefaultHttpRequest(HTTP_1_1, POST, "/form");
+    private static final HttpRequest req2 = new DefaultHttpRequest(HTTP_1_1, POST, "/form");
+
+    private DefaultHttpDataFactory factory;
+
+    @BeforeClass
+    public static void assertReq1EqualsReq2() {
+        // Before doing anything, assert that the requests are equal
+        assertEquals(req1.hashCode(), req2.hashCode());
+        assertTrue(req1.equals(req2));
+    }
+
+    @Before
+    public void setupFactory() {
+        factory = new DefaultHttpDataFactory();
+    }
+
+    @After
+    public void cleanupFactory() {
+        factory.cleanAllHttpData();
+    }
+
+    @Test
+    public void cleanRequestHttpDataShouldIdentifiesRequestsByTheirIdentities() throws Exception {
+        // Create some data belonging to req1 and req2
+        Attribute attribute1 = factory.createAttribute(req1, "attribute1", "value1");
+        Attribute attribute2 = factory.createAttribute(req2, "attribute2", "value2");
+        FileUpload file1 = factory.createFileUpload(
+                req1, "file1", "file1.txt",
+                DEFAULT_TEXT_CONTENT_TYPE, IDENTITY.toString(), UTF_8, 123
+        );
+        FileUpload file2 = factory.createFileUpload(
+                req2, "file2", "file2.txt",
+                DEFAULT_TEXT_CONTENT_TYPE, IDENTITY.toString(), UTF_8, 123
+        );
+        file1.setContent(Unpooled.copiedBuffer("file1 content", UTF_8));
+        file2.setContent(Unpooled.copiedBuffer("file2 content", UTF_8));
+
+        // Assert that they are not deleted
+        assertNotNull(attribute1.getByteBuf());
+        assertNotNull(attribute2.getByteBuf());
+        assertNotNull(file1.getByteBuf());
+        assertNotNull(file2.getByteBuf());
+
+        // Clean up by req1
+        factory.cleanRequestHttpData(req1);
+
+        // Assert that data belonging to req1 has been cleaned up, but not req2
+        assertNull(attribute1.getByteBuf());
+        assertNull(file1.getByteBuf());
+        assertNotNull(attribute2.getByteBuf());
+        assertNotNull(file2.getByteBuf());
+    }
```

Comment:
The reference counts here are not cleaned up ... but that maybe because of a bug in `DefaultHttpDataFactory`. Lets discuss on the previous comments.